packages/common/container-definitions/src/containerExtension.ts

```diff
@@ -0,0 +1,293 @@
+					 */
+					targetClientId?: ClientConnectionId;
+				}
+			>
+		: never;
+
+/**
+ * Outgoing extension signals.
+ *
+ * @sealed
+ * @internal
+ */
+export type OutboundExtensionMessage<TMessage extends TypedMessage = TypedMessage> =
+	ExtensionMessage<{ type: TMessage["type"]; content: JsonSerializable<TMessage["content"]> }>;
+
+/**
+ * Brand for value that has not been verified.
+ *
+ * Usage:
+ *
+ * - Cast to with `as unknown as UnverifiedBrand<T>` when value of or containing expected type `T` is yet unknown.
+ *
+ * - Cast from with `as unknown` when "instance" will be parsed to `T`.
+ *
+ * @sealed
+ * @internal
+ */
+export declare class UnverifiedBrand<T> extends BrandedType<T> {
+	private readonly UnverifiedValue: T;
+	private constructor();
+}
+
+/**
+ * Unverified incoming extension signals.
+ *
+ * @sealed
+ * @internal
+ */
+export type RawInboundExtensionMessage<TMessage extends TypedMessage = TypedMessage> =
+	// `TMessage extends TypedMessage` encourages processing union elements individually
+	TMessage extends TypedMessage
+		? InternalUtilityTypes.FlattenIntersection<
+				ExtensionMessage<{
+					type: string;
+					content: JsonDeserialized<unknown>;
+				}> & {
+					/**
+					 * The client ID that submitted the message.
+					 * For server generated messages the clientId will be null.
+					 */
+					// eslint-disable-next-line @rushstack/no-new-null
+					clientId: ClientConnectionId | null;
+				}
+			> &
+				UnverifiedBrand<TMessage>
+		: never;
+
+/**
+ * Verified incoming extension signals.
+ *
+ * @sealed
+ * @internal
+ */
+export type VerifiedInboundExtensionMessage<TMessage extends TypedMessage = TypedMessage> =
+	// `TMessage extends TypedMessage` encourages processing union elements individually
+	TMessage extends TypedMessage
+		? InternalUtilityTypes.FlattenIntersection<
+				ExtensionMessage<{
+					type: TMessage["type"];
+					content: JsonDeserialized<TMessage["content"]>;
+				}> & {
+					/**
+					 * The client ID that submitted the message.
+					 * For server generated messages the clientId will be null.
+					 */
+					// eslint-disable-next-line @rushstack/no-new-null
+					clientId: ClientConnectionId | null;
+				}
+			>
+		: never;
+
+/**
+ * Incoming extension signal that may be of the known type or has not yet been validated.
+ *
+ * @sealed
+ * @internal
+ */
+export type InboundExtensionMessage<TMessage extends TypedMessage = TypedMessage> =
+	| RawInboundExtensionMessage<TMessage>
+	| VerifiedInboundExtensionMessage<TMessage>;
+
+/**
+ * @internal
+ */
+export interface ExtensionRuntimeProperties {
+	SignalMessages: TypedMessage;
+}
+
+/**
+ * Defines requirements for a component to register with container as an extension.
+ *
+ * @internal
+ */
+export interface ContainerExtension<
+	TUseContext extends unknown[],
+	TRuntimeProperties extends ExtensionRuntimeProperties,
+> {
+	/**
+	 * Notifies the extension of a new use context.
```

Comment:
Let's have a beefier comment here explaining what a "new use context" is.  And if there's a way to rename it to avoid "context"...? :)